### PR TITLE
Assignment 5

### PR DESCRIPTION
**Pull request by JerettLatimer**

### CODE REVIEW

File: Mailbox/DataLoader.cs [hunk 1]
```diff
@@ -1,21 +1,73 @@
...
+            {
+                foreach (Mailbox mailbox in mailboxes)
+                {
+                    writer.WriteLine(JsonConvert.SerializeObject(mailbox));
+                }
+            }
         }
+
+        #region IDisposable Support
+        private bool disposedValue = false; // To detect redundant calls
```
Comment (Keboo) [inline Mailbox/DataLoader.cs]: Fields should be specified in _PascalCase, or switched to use an auto property

File: Mailbox/Program.cs [hunk 2]
```diff
@@ -69,7 +69,9 @@ static void Main(string[] args)
                     case 4:
```
Comment (Keboo) [inline Mailbox/Program.cs]: Parameter args of method Main is never used. Remove the parameter or use it in the method body.

File: Mailbox/Mailbox.cs [hunk 3]
```diff
@@ -1,4 +1,37 @@
...
+
+namespace Mailbox
 {
+    public class Mailbox
+    {
+        public Size Size { get; set; }
+        public (int x, int y) Location { get; set; }
+        public Person Owner { get; set; }
+
+        public Mailbox(Size size, ValueTuple<int,int> location, Person owner)
```
Comment (Keboo) [inline Mailbox/Mailbox.cs]: ValueTuple syntax can be simplified
```suggestion
        public Mailbox(Size size, (int,int) location, Person owner)
```

File: Mailbox.Tests/DataLoaderTests.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            list.Add(new Mailbox(Size.Large, (1, 6), new Person("first", "Last")));
+            list.Add(new Mailbox(Size.Large, (3, 1), new Person("first", "Last")));
+
+            // Act
+            sut.Save(list);
+            List<Mailbox> testList = sut.Load();
+
+            // Assert
+            Assert.AreEqual(testList.Count, list.Count);
+            sut.Dispose();
```
Comment (Keboo) [inline Mailbox.Tests/DataLoaderTests.cs]: You want to take advantage of the `using` for disposing objects. If the above code throws an exception (such as the case where an Assert fails) this `Dispose` will never get called.
```C#
using DataLoader sut = new DataLoader(source);
```

File: Mailbox.Tests/MailboxTests.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        public void Mailbox_ToString_ReturnsCorrectlyFormattedString(Size size)
+        {
+            // Arrange
+            Person testPerson = new Person("First", "Last");
+            Mailbox sut = new Mailbox(size, (1, 1), testPerson);
+
+            // Act
+
+            // Assert
+            Assert.AreEqual(sut.ToString(), $"Size: {size} Location: (1, 1) Owner: First Last");
```
Comment (Keboo) [inline Mailbox.Tests/MailboxTests.cs]: This "works" but your two parameters are backwards. The first parameter is the "expected" value and the second is the "actual" value. With them backwards the error message that is displayed on failure will be wrong.

File: Mailbox/DataLoader.cs [hunk 6]
```diff
@@ -1,21 +1,73 @@
...
 
 namespace Mailbox
 {
-    public class DataLoader
+    public class DataLoader : IDisposable
     {
+        private Stream Source { get; set; }
         public DataLoader(Stream source)
         {
+            Source = source;
```
Comment (Keboo) [inline Mailbox/DataLoader.cs]: `source` should be null checked
```suggestion
            Source = source ?? throw new ArgumentNullException(nameof(source));
```

File: Mailbox/Mailbox.cs [hunk 7]
```diff
@@ -1,4 +1,37 @@
...
+        public override string ToString()
+        {
+            String size;
+            if(Size == Size.Default)
+            {
+                size = "";
+            }
+            else if(Size == Size.Premium)
+            {
+                size = Size.ToString() + "Premium";
```
Comment (Keboo) [inline Mailbox/Mailbox.cs]:
```suggestion
                size = Size.ToString() + " Premium";
```

File: Mailbox/Person.cs [hunk 8]
```diff
@@ -1,4 +1,66 @@
...
+namespace Mailbox
 {
+#pragma warning disable CA1067 // Override Object.Equals(object) when implementing IEquatable<T>
+    public struct Person : IEquatable<Person>
+#pragma warning restore CA1067 // Override Object.Equals(object) when implementing IEquatable<T>
+    {
+        private string? FirstName { get; set; }
+        private string? LastName { get; set; }
+
+        public Person(String? firstName, String? lastName)
```
Comment (Keboo) [inline Mailbox/Person.cs]:
```suggestion
        public Person(string? firstName, string? lastName)
```

File: Mailbox/Person.cs [hunk 9]
```diff
@@ -1,4 +1,66 @@
...
+using System.Diagnostics.CodeAnalysis;
+
+namespace Mailbox
 {
+#pragma warning disable CA1067 // Override Object.Equals(object) when implementing IEquatable<T>
+    public struct Person : IEquatable<Person>
+#pragma warning restore CA1067 // Override Object.Equals(object) when implementing IEquatable<T>
+    {
+        private string? FirstName { get; set; }
+        private string? LastName { get; set; }
```
Comment (Keboo) [inline Mailbox/Person.cs]: Given the null checks in the constructor these could be made readonly non-nullable properties.
```suggestion
        private string LastName { get; }
```

File: Mailbox/Person.cs [hunk 10]
```diff
@@ -1,4 +1,66 @@
...
+            }
+            else if(lastName is null)
+            {
+                throw new ArgumentNullException(nameof(lastName));
+            }
+            FirstName = firstName;
+            LastName = lastName;
+        }
+
+        public bool Equals(Person other)
```
Comment (Keboo) [inline Mailbox/Person.cs]: `other` should be null checked

File: Mailbox/Person.cs [hunk 11]
```diff
@@ -1,4 +1,66 @@
...
+        public override string ToString()
+        {
+            return $"{FirstName} {LastName}";
+        }
+
+        public override bool Equals(object? obj)
+        {
+            if (obj is null)
+                return false;
+            return Equals((Person)obj);
```
Comment (Keboo) [inline Mailbox/Person.cs]: This will cause an invalid cast exception if this is compared with something that is not a Person.
```suggestion
            return obj is Person person && Equals(person);
```

File: Mailbox/Program.cs [hunk 12]
```diff
@@ -89,17 +91,41 @@ static void Main(string[] args)
 
         public static string GetOwnersDisplay(Mailboxes mailboxes)
         {
-            
+            string owners = "";
+            foreach(Mailbox mailbox in mailboxes)
+            {
+                owners += mailbox.Owner.ToString() + " ";
+            }
+            return owners;
```
Comment (Keboo) [inline Mailbox/Program.cs]: The expectation on was that this method return a _distinct_ set of owner names. This just returns all of them.
